src/fluree/db/api/query.cljc

```diff
@@ -152,183 +157,110 @@
-                                                                       (flake/flip-flake flake))]
-                                           (if asserted?
-                                             (recur r (update asserted-subjects subject #(vec (conj % flake')))
-                                                    retracted-subjects)
-                                             (recur r asserted-subjects
-                                                    (update retracted-subjects subject #(vec (conj % flake'))))))))
-                retracted (loop [[subject & r] (vals retracted-subjects)
-                                 acc []]
-                            (if-not subject
-                              acc
-                              (recur r (conj acc (<? (flakes->res db cache fuel 1000000 {:wildcard? true, :select {}}
-                                                                  {} subject))))))
-                asserted  (loop [[subject & r] (vals asserted-subjects)
-                                 acc []]
-                            (if-not subject
-                              acc
-                              (recur r (conj acc (<? (flakes->res db cache fuel 1000000 {:wildcard? true, :select {}}
-                                                                  {} subject))))))]
-            {:block     (:block curr-block)
-             :t         (:t curr-block)
-             :retracted retracted
-             :asserted  asserted})))
-
-
-(defn- format-blocks-resp-pretty
-  [db resp]
-  (async/go-loop [fuel (volatile! 0)
-                  cache (volatile! {})
-                  curr-block (first resp)
-                  rest-blocks (rest resp)
-                  acc []]
-    (let [curr-block' (<? (format-block-resp-pretty db curr-block cache fuel))
-          acc'        (concat acc [curr-block'])]
-      (if (first rest-blocks)
-        (recur fuel cache (first rest-blocks) (rest rest-blocks) acc')
-        acc'))))
-
+(defn history-flakes->json-ld
+  [db q flakes]
+  (go-try
+    (let [fuel    (volatile! 0)
+          cache   (volatile! {})
+          compact (json-ld/compact-fn (fql-parse/parse-context q db))]
+      (->> (sort-by flake/t flakes)
+           (partition-by flake/t)
+           (map (fn [t-flakes]
+                  (go-try
+                    (let [assert-flakes (not-empty (filter flake/op t-flakes))
+                          retract-flakes (not-empty (filter (complement flake/op) t-flakes))
+
+                          asserts (<? (json-ld-resp/flakes->res db cache compact fuel 1000000
+                                                                {:wildcard? true, :depth 0} 0 assert-flakes))
+                          retracts (<? (json-ld-resp/flakes->res db cache compact fuel 1000000
+                                                                 {:wildcard? true, :depth 0}
+                                                                 0 retract-flakes))]
```

Comment:
Does the `pipeline-async` usage achieve some of these performance goals? I'm not using a `take!`, but I'm not really sure what the `fn` `take!` takes would do in this context. I'm still a bit of a core.async neophyte so composing these async operations is still not always clear.

Reply:
<? blocks so your retracts var cannot start until asserts is finished. Since they have no dependency on each other they could be done in parallel just by moving your blocking takes (<?) later in the code. If this is all inside a pipeline and the pipeline is using all cores this has no gains as you’ve maxed out parallelism already - but unlikely unless server is under load.